### ISSUE TITLE
Complete Esim implementation

### ISSUE THREAD

**Issue by kalyangvs** [coetaur0/ESIM#1]
Can you please state the reason as to why HIM (Hybrid Inference Model) is not implemented in many ESIM implementations.
Is their no visible improvement when Tree LSTM is added ?
which is advisable plain ESIM or HIM (considering the time to get an inference too)
or is BiMPM better of the three?

**Comment by coetaur0** [comment]
Hi, it's a good question, actually. I don't really know, but I can try to answer.

I think one of the reasons why the ESIM model is still so popular even though it's not state of the art anymore is because it is rather easy to understand and fast to train. HIM's inner workings might be a bit harder to grasp because of the TreeLSTMs it uses, which I guess most people aren't too familiar with.

The reported performance gain of HIM over ESIM is 0.6% accuracy. I haven't tried training and testing HIM, but since LSTMs are the slowest component to execute in ESIM, I assume that including TreeLSTMs in the model would increase its training and testing times quite significantly.
I cannot really speak for other implementations, but I imagine the additional complexity and slower execution times of HIM don't seem worth the small accuracy increase to most people who implement the model. In my case, the main reason why I didn't implement HIM was because I only needed a version of ESIM for some experiments.

There's a nice summary of all existing models' performances on SNLI on the [official website](https://nlp.stanford.edu/projects/snli/) of the dataset. BiMPM's reported accuracy is a little lower than ESIM and HIM's. The current SOTA reported there is the Fine-Tuned LM-Pretrained Transformer by Radford et al., but Google recently released [BERT](https://arxiv.org/abs/1810.04805), another transformer model that outperforms all existing systems.

**Comment by kalyangvs** [comment]
Thanks.
